eth/stagedsync/stage_headers.go

```diff
@@ -162,10 +164,97 @@ func HeadersDownward(
 		return err
 	}
 	if parent != nil && parent.Hash() == header.ParentHash {
-		return s.Update(tx, header.Number.Uint64())
+		if err := s.Update(tx, header.Number.Uint64()); err != nil {
+			return err
+		}
+		// For the sake of simplicity we can just assume it will be valid for now. (TODO: move to execution stage)
+		cfg.statusCh <- privateapi.ExecutionStatus{
+			Status:   privateapi.Valid,
+			HeadHash: header.Hash(),
+		}
+		return tx.Commit()
+	}
+	cfg.hd.SetBackwards(true)
+	if err = cfg.hd.ReadProgressFromDb(tx); err != nil {
+		return err
+	}
+	cfg.hd.SetProcessed(header.Number.Uint64())
+	cfg.hd.SetExpectedHash(header.ParentHash)
+	cfg.hd.SetFetching(true)
+	defer cfg.hd.SetFetching(false)
+	logPrefix := s.LogPrefix()
+
+	logEvery := time.NewTicker(logInterval)
+	defer logEvery.Stop()
+
+	// Allow other stages to run 1 cycle if no network available
+	if initialCycle && cfg.noP2PDiscovery {
+		return nil
+	}
+
+	cfg.statusCh <- privateapi.ExecutionStatus{
+		Status:   privateapi.Syncing,
+		HeadHash: rawdb.ReadHeadBlockHash(tx),
 	}
-	// Downward sync if we need to process more (TODO)
-	return s.Update(tx, header.Number.Uint64())
+	log.Info(fmt.Sprintf("[%s] Waiting for headers...", logPrefix), "from", header.Number.Uint64())
+
+	cfg.hd.SetHeaderReader(&chainReader{config: &cfg.chainConfig, tx: tx, blockReader: cfg.blockReader})
+
+	stopped := false
+	prevProgress := header.Number.Uint64()
+	go func() {
+		for {
+			timer := time.NewTimer(1 * time.Second)
+			select {
+			case <-ctx.Done():
+				stopped = true
+			case <-logEvery.C:
+				diff := prevProgress - cfg.hd.Progress()
+				if cfg.hd.Progress() <= prevProgress {
+					log.Info("Wrote Block Headers backwards", "from", header.Number.Uint64(),
+						"now", cfg.hd.Progress(), "blk/sec", float64(diff)/float64(logInterval/time.Second))
+					prevProgress = cfg.hd.Progress()
+				}
+			}
+			timer.Stop()
+		}
+	}()
+	headerCollector := etl.NewCollector(logPrefix, cfg.tmpdir, etl.NewSortableBuffer(etl.BufferOptimalSize))
+	canonicalHeadersCollector := etl.NewCollector(logPrefix, cfg.tmpdir, etl.NewSortableBuffer(etl.BufferOptimalSize))
+	cfg.hd.SetHeadersCollector(headerCollector)
+	cfg.hd.SetCanonicalHashesCollector(canonicalHeadersCollector)
+	defer headerCollector.Close()
+	defer canonicalHeadersCollector.Close()
+	var req headerdownload.HeaderRequest
+	for !stopped {
```

Comment:
`stopped` never set to `true` inside this loop. Also, this looks like a "tight loop", meaning that it will churn out requests and try to push them to the sentry, without any pauses, this will eat CPU and also put a lot of pressure on the communication with the sentry. You need some kind of back off logic for the case when `sentToPeer` is false, to let peers respond to previous requests. Look how it is done for forward sync, I believe there is 1 second timer

Reply:
You can also check out in the forward sync how `ctx.Done()` is used to set `stopped = false` so that this process can be gracefully broken by Ctrl-C or by sending `SIGTERM` signal (which will lead to closing of the context `ctx`). And such check also need to happen within sub loop below

Reply:
i do it in the goroutine, now i also check it in the request loop

Reply:
It is usually not advised to use a variable from more than one goroutine without any synchronisation. Even if the variable can be atomically changed. Because of any possible caching. You need to either use mutexes, or atomics, or channels, or remove that extra goroutine.

Reply:
now i am using channels

Reply:
Ok, that should work